### PR TITLE
support defaultValue for input types

### PR DESCRIPTION
**Pull request by benawad**
fix: https://github.com/19majkel94/type-graphql/issues/53
improved on: https://github.com/19majkel94/type-graphql/pull/141
I'm happy to iterate on this further if any changes are needed.

### CODE REVIEW

**Comment by 19majkel94** [comment]
Looks ok but it's just a beginning. 
Please take a look at this comment:
https://github.com/19majkel94/type-graphql/pull/141#pullrequestreview-151603632

File: tests/functional/resolvers.ts [hunk 1]
```diff
@@ -104,6 +104,7 @@ describe("Resolvers", () => {
           @Arg("stringArrayArg", type => String) stringArrayArg: string[],
           @Arg("explicitArrayArg", type => [String]) explicitArrayArg: any,
           @Arg("nullableStringArg", { nullable: true }) nullableStringArg?: string,
+          @Arg("defaultStringArg", { defaultValue: "bob" }) defaultStringArg?: string,
```
Comment (19majkel94) [inline tests/functional/resolvers.ts]: Please try to follow the descriptive names convention, like `defaultStringArgDefaultValue` instead of `foo`, `bar`, `bob` and other sample names that might generate conflicts.

**Comment by benawad** [comment]
What is `Support for initializers in classes ("reflection") - creating an instance of the class and checking the value of the properties to detect default values`?

**Comment by 19majkel94** [comment]
@benawad 
```ts
@InputType()
class SampleInput {
  @Field()
  defaultField: string = "defaultFieldValue"
}
```
```graphql
input SampleInput {
  defaultField: String = "defaultFieldValue"
}
```

**Comment by benawad** [comment]
How should I go about doing that?

**Comment by 19majkel94** [comment]
@benawad 
Just by "creating an instance of the class and checking the value of the properties to detect default values" 😃 

```ts
class SampleInput {
  normalField: string;
  stringFieldWithDefault: string = "defaultValue"
}

const instance = new SampleInput();

const defaultsMap = Object.keys(instance).map(field => ({
  field,
  defaultValue: instance[field],
}));

console.log(defaultsMap);
```

**Comment by benawad** [comment]
I was able to get it working for input objects, but I'm having trouble figuring out how to do it with args

**Comment by 19majkel94** [comment]
@benawad 
You need to do the same things as with input - just create the instance in `mapArgFields`, read the defaults and attach them to `args[field.schemaName]` config:
```ts
const instance = new (argumentType.target as any)();
```

File: tests/functional/resolvers.ts [hunk 2]
```diff
@@ -104,6 +114,10 @@ describe("Resolvers", () => {
           @Arg("stringArrayArg", type => String) stringArrayArg: string[],
           @Arg("explicitArrayArg", type => [String]) explicitArrayArg: any,
           @Arg("nullableStringArg", { nullable: true }) nullableStringArg?: string,
+          @Arg("defaultStringArg", { defaultValue: "defaultStringArgDefaultValue" })
```
Comment (19majkel94) [inline tests/functional/resolvers.ts]: Object types have no default values, only input types can have by GraphQL specification

**Comment by 19majkel94** [comment]
We should also test some edge cases, like overwriting default values in child classes (implicit and by decorator option):
```ts
class Base {
  prop = "baseProp"
}

class Child extends Base {
  prop = "childProp"
}
```

And I think that we should report to user when `implicitDefaultValue` has the value and `field.typeOptions.defaultValue` has the value and they're not equal to each other.
Now it's just silently overwrites the default value from decorator.

**Comment by benawad** [comment]
Is it possible to detect the default values for function parameters?
```
          @Arg("implicitDefaultStringArg")
          implicitDefaultStringArg: string = "implicitDefaultStringArgDefaultValue",
```

**Comment by benawad** [comment]
Would you like me to make any other changes?

File: tests/functional/resolvers.ts [hunk 3]
```diff
@@ -224,6 +246,12 @@ describe("Resolvers", () => {
         type => type.name === "SampleObject",
       ) as IntrospectionObjectType;
       argMethodField = sampleObjectType.fields.find(field => field.name === "argMethodField")!;
+      sampleInputType = schemaIntrospection.types.find(
```
Comment (19majkel94) [inline tests/functional/resolvers.ts]: This variables are not used globally in this test suite - move them do the used block:
```ts
describe("Input object", () => {
  let sampleInputType: IntrospectionInputObjectType;
  let sampleInputChildType: IntrospectionInputObjectType;

  beforeAll(() => {
    sampleInputType = schemaIntrospection.types.find(
      field => field.name === "SampleInput",
    )! as IntrospectionInputObjectType;
    sampleInputChildType = schemaIntrospection.types.find(
      field => field.name === "SampleInputChild",
    )! as IntrospectionInputObjectType;
  });
```

File: tests/functional/resolvers.ts [hunk 4]
```diff
@@ -100,10 +119,13 @@ describe("Resolvers", () => {
           @Arg("booleanArg") booleanArg: boolean,
           @Arg("numberArg") numberArg: number,
           @Arg("inputArg") inputArg: SampleInput,
+          @Arg("inputChildArg") inputChildArg: SampleInputChild,
           @Arg("explicitNullableArg", type => String, { nullable: true }) explicitNullableArg: any,
           @Arg("stringArrayArg", type => String) stringArrayArg: string[],
           @Arg("explicitArrayArg", type => [String]) explicitArrayArg: any,
           @Arg("nullableStringArg", { nullable: true }) nullableStringArg?: string,
+          @Arg("defaultStringArg", { defaultValue: "defaultStringArgDefaultValue" })
+          defaultStringArg?: string,
```
Comment (19majkel94) [inline tests/functional/resolvers.ts]: It has `defaultValue` so it's not optional (`?`) anymore

File: tests/functional/resolvers.ts [hunk 5]
```diff
@@ -74,6 +89,10 @@ describe("Resolvers", () => {
         numberArg: number;
         @Field()
         inputObjectArg: SampleInput;
+        @Field({ defaultValue: "defaultStringArgDefaultValue" })
+        defaultStringArg?: string;
```
Comment (19majkel94) [inline tests/functional/resolvers.ts]: The same - no optional when defaults used

File: tests/functional/resolvers.ts [hunk 6]
```diff
@@ -74,6 +89,10 @@ describe("Resolvers", () => {
...
+        @Field()
+        implicitDefaultStringArg?: string = "implicitDefaultStringArgDefaultValue";
```
Comment (19majkel94) [inline tests/functional/resolvers.ts]: No optional when defaults used

File: tests/functional/resolvers.ts [hunk 7]
```diff
@@ -375,6 +403,92 @@ describe("Resolvers", () => {
         expect(inputArgInnerType.kind).toEqual(TypeKind.INPUT_OBJECT);
         expect(inputArgInnerType.name).toEqual("SampleInput");
       });
+
+      it("should generate nullable string arg type with defaultValue for object field method", async () => {
+        const inputArg = argMethodField.args.find(arg => arg.name === "defaultStringArg")!;
+        const defaultValueStringArgType = inputArg.type as IntrospectionNamedTypeRef;
+
+        expect(inputArg.defaultValue).toBe('"defaultStringArgDefaultValue"');
```
[the remark this answers is not in the document]
Comment (19majkel94) [reply tests/functional/resolvers.ts]: Ok, it looks like it just works like that. Weird:
```
        {
          "name": "ratingsCount",
          "args": [
            {
              "name": "minRate",
              "defaultValue": "0"
            }
          ]
        },
```
So it has to stay like this 😕

**Comment by 19majkel94** [comment]
@benawad 
> Would you like me to make any other changes?

Yes, we need to add info about all this things in docs 😃 
https://19majkel94.github.io/type-graphql/docs/resolvers.html
- `defaultValue` in `@Arg` decorator
- `defaultValue` in `@Field` decorator (only for `@InputType` and `@ArgsType`)
- initialization of property affects default value in schema
- inheritance of default values
- warning about conflict between decorator and and initializer default value

Also, examples and docs might need some updates when there's a default value together with `{ nullable: true }` decorator's option.

And how do we handle the edge case when parent class has property initializer but child try to overwrites it with decorator option? Does it throw error?
And what about when parent has decorator option and child overwrites it with property initalizer?
Do we even test the simple inheritance with no overwrites? 😕

**Comment by benawad** [comment]
"And how do we handle the edge case when parent class has property initializer but child try to overwrites it with decorator option? Does it throw error?"
"And what about when parent has decorator option and child overwrites it with property initalizer?"

I think we should take the child defaultValue in both cases without throwing an error because they could be inheriting from a class from a library or something and not know (and don't care) how thaat class set the default value. 

"Do we even test the simple inheritance with no overwrites? 😕"

Nope, and I just added a test for it and it failed 😂. Where in the code do you handle inheritance?

I'll start working on the docs next.

**Comment by 19majkel94** [comment]
> Where in the code do you handle inheritance?

Before #183 it's all over the place 😆
Now I've realized that `@Authorized` also doesn't work with inheritance 😞 	

Args classes inheritance is here:
https://github.com/benawad/type-graphql/blob/d7659fb0d446d3a35ba7943b7f82a60bb939d3bb/src/schema/schema-generator.ts#L453-L460

Input types inheritance is here:
https://github.com/benawad/type-graphql/blob/d7659fb0d446d3a35ba7943b7f82a60bb939d3bb/src/schema/schema-generator.ts#L330-L336

> I think we should take the child defaultValue in both cases

Right! Make sure that the tests are covering this cases 😉 	

Thanks and keep going! 💪

**Comment by 19majkel94** [comment]
@benawad
> start working on docs

Are you going to finish this up or should I checkout your branch and add some changes there?

**Comment by benawad** [comment]
I'm not sure when I'll get time to finish this up, so it would be great if you want to.

**Comment by 19majkel94** [comment]
@benawad 
I think that we are ready to go with this feature.
Can you take a look at my changes and review them? 😉

**Comment by benawad** [comment]
I'll take a look at it tonight

**Comment by benawad** [comment]
Looks good
